### PR TITLE
Add aten::erfinv, aten::exp2, aten::expm1, aten::exponential_

### PR DESCRIPTION
**Pull request by hjhee**
- erfinv
  - erfinv_
  - erfinv.out
  - exp2
  - exp2_
  - exp2.out
  - expm1
  - expm1_
  - expm1.out
  - exponential_

### CODE REVIEW

File: src/ATen/native/xpu/sycl/DistributionTemplates.h [hunk 1]
```diff
@@ -590,6 +590,51 @@ void bernoulli_kernel(TensorIteratorBase& iter, double p, RNG gen) {
...
+  }
+  ExponentialFunctor(accscalar_t lambd) : lambd_(lambd) {}
+
+ private:
+  accscalar_t lambd_;
+};
+
+template <typename RNG>
+void exponential_kernel(TensorIteratorBase& iter, double lambda, RNG gen) {
+  AT_DISPATCH_ALL_TYPES_AND3(
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/DistributionTemplates.h:623]: Please align the dtypes.
image

File: src/ATen/native/xpu/sycl/DistributionTemplates.h [hunk 2]
```diff
@@ -591,6 +591,54 @@ void bernoulli_kernel(TensorIteratorBase& iter, double p, RNG gen) {
...
+    }
+    return static_cast<accscalar_t>(-1.f) / lambd_ * log;
+  }
+  ExponentialFunctor(accscalar_t lambd) : lambd_(lambd) {}
+
+ private:
+  accscalar_t lambd_;
+};
+
+template <typename RNG>
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/DistributionTemplates.h:622]: Should not a template function. Nothing will trigger the instantiation of the template.

Comment (fengyuan14) [reply src/ATen/native/xpu/sycl/DistributionTemplates.h:622]: You need hard-code the type of generator by XPU generator type.

File: src/ATen/native/xpu/sycl/DistributionTemplates.h [hunk 3]
```diff
@@ -591,6 +591,54 @@ void bernoulli_kernel(TensorIteratorBase& iter, double p, RNG gen) {
...
+    // https://github.com/pytorch/pytorch/issues/16706
+    // rand_uniform has (0,1] bounds. log(1) is 0 and exponential
+    // excludes 0. we need log to be not 0, and not underflow when
+    // converted to half
+    accscalar_t log;
+    if (val >= static_cast<accscalar_t>(1.f) -
+            std::numeric_limits<scalar_t>::epsilon() / 2.f) {
+      // Need an epsilon of appropriate precision.
+      // Unlike CUDA behavior, DPCPP dtype conversions do not support
+      // epsilon downgrading.
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/DistributionTemplates.h:609]: Remove the comments.

File: src/ATen/native/xpu/XPUFallback.template
Comment (fengyuan14) [inline src/ATen/native/xpu/XPUFallback.template:1]: Align with CUDA `DistributionExponentialKernel.cpp`

File: src/ATen/native/xpu/sycl/UnarySpecialOpsKernels.cpp [hunk 4]
```diff
@@ -74,4 +74,130 @@ void erfc_kernel(TensorIteratorBase& iter) {
...
+
+void expm1_kernel(TensorIteratorBase& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      iter.common_dtype(),
+      "expm1_xpu",
+      [&]() { gpu_kernel(iter, Expm1Functor<scalar_t>()); });
+}
+
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UnarySpecialOpsKernels.cpp:202]: Move to UnaryKernels.cpp

File: src/ATen/native/xpu/sycl/UnarySpecialOpsKernels.h [hunk 5]
```diff
@@ -10,4 +10,10 @@ void erf_kernel(TensorIteratorBase& iter);
 
 void erfc_kernel(TensorIteratorBase& iter);
 
+void erfinv_kernel(TensorIteratorBase& iter);
+
+void exp2_kernel(TensorIteratorBase& iter);
+
+void expm1_kernel(TensorIteratorBase& iter);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UnarySpecialOpsKernels.h:17]: Move to UnaryKernels.h

**Comment by fengyuan14** [comment]
Please check the failure. Most likely we have different handle between GCC and SYCL compiler in std::exp2 for std::complex.
image

**Comment by hjhee** [comment]
Handling logic for CPU/XPU is not consistent for `exp2`:
```python
cpu_results
tensor([nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj])
cuda_results
tensor([inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj,
        inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, 0.+0.j,
        0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j,
        0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
        nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj],
       device='xpu:0')
cpu_sample.input
tensor([inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf-infj, inf-infj, inf-infj, inf-infj, inf-infj,
        inf-infj, inf-infj, inf-infj, inf-infj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, -inf+infj,
        -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf-infj, -inf-infj, -inf-infj, -inf-infj, -inf-infj, -inf-infj,
        -inf-infj, -inf-infj, -inf-infj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, nan+infj, nan+infj,
        nan+infj, nan+infj, nan+infj, nan+infj, nan+infj, nan+infj, nan+infj, nan-infj, nan-infj, nan-infj, nan-infj, nan-infj, nan-infj, nan-infj,
        nan-infj, nan-infj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj])
```

**Comment by fengyuan14** [comment]
> Handling logic for CPU/XPU is not consistent for `exp2`:
> 
> ```python
> cpu_results
> tensor([nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj])
> cuda_results
> tensor([inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj,
>         inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, 0.+0.j,
>         0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j,
>         0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj,
>         nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj],
>        device='xpu:0')
> cpu_sample.input
> tensor([inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf+infj, inf-infj, inf-infj, inf-infj, inf-infj, inf-infj,
>         inf-infj, inf-infj, inf-infj, inf-infj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, inf+nanj, -inf+infj,
>         -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf+infj, -inf-infj, -inf-infj, -inf-infj, -inf-infj, -inf-infj, -inf-infj,
>         -inf-infj, -inf-infj, -inf-infj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, -inf+nanj, nan+infj, nan+infj,
>         nan+infj, nan+infj, nan+infj, nan+infj, nan+infj, nan+infj, nan+infj, nan-infj, nan-infj, nan-infj, nan-infj, nan-infj, nan-infj, nan-infj,
>         nan-infj, nan-infj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj, nan+nanj])
> ```

We have met such kind of case in other std operators. We have different behavior for std::complex, which contains denormal value (nan or inf).

**Comment by fengyuan14** [comment]
When input is inf+infj, as the C++ standard, the output should be +-inf+nanj. XPU result should align with standard, but CPU result gets nan+nanj.
image

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**